linux/src/threadpool_linux.c

```diff
@@ -0,0 +1,303 @@
+// Copyright (C) Microsoft Corporation. All rights reserved.
+
+#include <stdlib.h>
+#include <inttypes.h>
+#include <stdio.h>
+#include <stdbool.h>
+
+#include "macro_utils/macro_utils.h"
+
+#include "c_logging/xlogging.h"
+
+#include "c_pal/threadapi.h"
+#include "c_pal/interlocked.h"
+#include "c_pal/sync.h"
+#include "c_pal/srw_lock.h"
+#include "c_pal/execution_engine.h"
+
+#include "c_pal/threadpool.h"
+
+#define TIMEOUT_MS      100
+#define MAX_THREAD_COUNT    8
+
+#define POOL_STATE_VALUES \
+    POOL_STATE_UNINIT, \
+    POOL_STATE_ACTIVE, \
+    POOL_STATE_IDLE
+
+MU_DEFINE_ENUM(POOL_STATE, POOL_STATE_VALUES)
+
+typedef struct THREADPOOL_TASK_TAG
+{
+    THREADPOOL_WORK_FUNCTION task_func;
+    void* task_param;
+    struct THREADPOOL_TASK_TAG* next;
+} THREADPOOL_TASK;
+
+typedef struct THREADPOOL_TAG
+{
+    volatile_atomic int32_t thread_count;
+    uint32_t max_thread_count;
+    volatile_atomic int32_t pool_state;
+    volatile_atomic int32_t quitting;
+    volatile_atomic int32_t task_count;
+
+    // Pool Queue
+    THREADPOOL_TASK* task_list;
+    THREADPOOL_TASK* task_list_last;
+
+    SRW_LOCK_HANDLE srw_lock;
+
+    THREAD_HANDLE* thread_handle_list;
+    uint32_t list_index;
+} THREADPOOL;
+
+int threadpool_work_func(void* param)
+{
+    if (param == NULL)
+    {
+        LogError("Failure unknown user parameter encountered");
+    }
+    else
+    {
+        THREADPOOL* threadpool = (THREADPOOL*)param;
+        do {
```

Comment:
> { [](http://example.com/codeflow?start=10&length=2)

Can we use K&R indenting? Preeeeety please?

Reply:
VS code does this automatically, and it's irritating